### DISCUSSION TITLE
Issue with iCub in Gazebo in tutorial_cartesian-interface

### DISCUSSION THREAD

**Discussion by paliasgh** [robotology/community#644] [Q&A] [answered]
Hello,

I am trying to run [tutorial_cartesian-interface](https://github.com/vvv-school/tutorial_cartesian-interface) on our updated system (robotology-superbuild v2023.02.3). `iKinCartesianSolver --context gazeboCartesianControl --part left_arm` in 'Tutorial on Cartesian Interface System' crashes shortly after start due to some errors (log file is attached). It tries to connect to some ports under the name `/icub`, instead of `/icubSim`. Everything else works fine. Could you please have a look into this issue?

[yarprunlog_24_07_2023_16_03_12.log](https://github.com/robotology/community/files/12149651/yarprunlog_24_07_2023_16_03_12.log)

**Comment by pattacini** [comment]
Hi @paliasgh 

I've tested the tutorial on fresh Gitpod-enabled systems (see #459):
- Against `robotology-superbuild@v2023.05.2` → https://gitpod.io/#https://github.com/vvv-school/tutorial_cartesian-interface.
- Against `robotology-superbuild@v2023.02.2`[^1] → https://gitpod.io/#https://github.com/vvv-school/tutorial_cartesian-interface/tree/gitpod-v2023.02.2.

Both are working as expected. This leads me to think that it may be due to some local changes to the files, perhaps. 

You can first pin down the actual configuration file that `iKinCartesianSolver` is using when launched by doing:
```console
yarp resource --context gazeboCartesianControl --from cartesianSolver.ini
```
Once found, you can inspect its content to check whether the stem names are consistent with `icubSim`.

It may also be that other copies of `cartesianSolver.ini` are found in directories different from the specified context, all the way down to the paths' hierarchy, which instead are intended to be used alongside the physical robot. If this is the case, it means that, for any reason, there's no `cartesianSolver.ini` file inside the `gazeboCartesianControl` build|install|.local folder(s).

[^1]: The only difference wrt `v2023.02.3` is the update of `robots-configuration` from `v2.2.1` to `v2.2.2`, hence nothing relevant to the tutorial.

**Reply by paliasgh** [reply]
Hello @pattacini ,

Thank you very much for getting back to me. You are right! Our system finds`cartesianSolver.ini` in another context, i.e., for our real robot, even if `gazeboCartesianControl` is specified:

```
icub@icub26:~$ yarp resource --context gazeboCartesianControl --from cartesianSolver.ini
[DEBUG] |yarp.os.ResourceFinder| configuring 
[DEBUG] |yarp.os.ResourceFinder| added context gazeboCartesianControl 
[DEBUG] |yarp.os.ResourceFinder| default config file specified as cartesianSolver.ini 
[DEBUG] |yarp.os.ResourceFinder| checking [/home/icub/cartesianSolver.ini] (pwd) 
[DEBUG] |yarp.os.ResourceFinder| checking [/home/icub/.config/yarp/robots/iCubWaterloo01] (robot YARP_CONFIG_HOME) 
[DEBUG] |yarp.os.ResourceFinder| checking [/home/icub/.local/share/yarp/robots/iCubWaterloo01] (robot YARP_DATA_HOME) 
[DEBUG] |yarp.os.ResourceFinder| found /home/icub/.local/share/yarp/robots/iCubWaterloo01 
[DEBUG] |yarp.os.ResourceFinder| checking [/etc/xdg/xdg-ubuntu/yarp/robots/iCubWaterloo01] (robot YARP_CONFIG_DIRS) 
[DEBUG] |yarp.os.ResourceFinder| checking [/etc/xdg/yarp/robots/iCubWaterloo01] (robot YARP_CONFIG_DIRS) 
[DEBUG] |yarp.os.ResourceFinder| checking [/usr/local/src/robot/robotology-superbuild/build/install/share/yarp/robots/iCubWaterloo01] (robot YARP_DATA_DIRS) 
[DEBUG] |yarp.os.ResourceFinder| checking [/usr/local/src/robot/robotology-superbuild/build/install/share/iCub/robots/iCubWaterloo01] (robot YARP_DATA_DIRS) 
[DEBUG] |yarp.os.ResourceFinder| checking [/usr/local/src/robot/robotology-superbuild/build/install/share/ICUBcontrib/robots/iCubWaterloo01] (robot YARP_DATA_DIRS) 
[DEBUG] |yarp.os.ResourceFinder| found /usr/local/src/robot/robotology-superbuild/build/install/share/ICUBcontrib/robots/iCubWaterloo01 
[DEBUG] |yarp.os.ResourceFinder| checking [/usr/local/src/robot/robotology-superbuild/build/install/share/ergoCub/robots/iCubWaterloo01] (robot YARP_DATA_DIRS) 
[DEBUG] |yarp.os.ResourceFinder| checking [/usr/local/src/robot/robotology-superbuild/build/install/share/yarp/config/path.d] (robot path.d YARP_DATA_DIRS) 
[DEBUG] |yarp.os.ResourceFinder| found /usr/local/src/robot/robotology-superbuild/build/install/share/yarp/config/path.d 
[DEBUG] |yarp.os.ResourceFinder| checking [/usr/local/src/robot/robotology-superbuild/build/install/share/speech/robots/iCubWaterloo01] (robot yarp.d) 
[DEBUG] |yarp.os.ResourceFinder| checking [/home/icub/.local/share/yarp/robots/iCubWaterloo01/cartesianSolver.ini] (robot) 
[DEBUG] |yarp.os.ResourceFinder| found /home/icub/.local/share/yarp/robots/iCubWaterloo01/cartesianSolver.ini 
[DEBUG] |yarp.os.ResourceFinder| finding file [from] 
"/home/icub/.local/share/yarp/robots/iCubWaterloo01/cartesianSolver.ini" 
```

However, we have the right `cartesianSolver.ini` inside `<superbuild-install-dir>/share/iCub/contexts/gazeboCartesianControl`, which is not being checked by `yarp resource`. I also saw a similar discussion [here](https://github.com/orgs/robotology/discussions/540) with your response to a quite similar issue. What do you suggest in our case, to let us run Gazebo simulations occasionally on a system which is used to work with our real robot?

**Reply by pattacini** [reply, marked as answer]
Hi @paliasgh 

>If this is the case, it means that, for any reason, there's no cartesianSolver.ini file inside the gazeboCartesianControl build|install|.local folder(s).

Yes, robot paths always override standard context paths. I forgot about this case.
This is for a good reason as it serves to enable robot-specific customizations.

My suggestion is to customize the name of the file in question, which is now shared among the robot and Gazebo.

```console
yarp-config context --import gazeboCartesianControl cartesianSolver.ini
cd ~/.local/share/yarp/contexts/gazeboCartesianControl
mv cartesianSolver.ini gazeboCartesianSolver.ini
```

Check if the file gets found:
```console
yarp resource --context gazeboCartesianControl --from gazeboCartesianSolver.ini
```

This way, you can configure `iKinCartesianSolver` correctly:
```console
iKinCartesianSolver --context gazeboCartesianControl --from gazeboCartesianSolver.ini --part left_arm
```

Let me know if that helps.

**Reply by pattacini** [reply]
Read more about contexts at #42.

**Reply by paliasgh** [reply]
Thank you very much @pattacini ! it worked.

**Reply by pattacini** [reply]
Great!
